tools/lambda-promtail/lambda-promtail/promtail.go

```diff
@@ -0,0 +1,166 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	timeout    = 5 * time.Second
+	minBackoff = 100 * time.Millisecond
+	maxBackoff = 30 * time.Second
+	maxRetries = 10
+
+	reservedLabelTenantID = "__tenant_id__"
+
+	userAgent = "lambda-promtail"
+)
+
+type entry struct {
+	labels model.LabelSet
+	entry  logproto.Entry
+}
+
+type batch struct {
+	streams map[string]*logproto.Stream
+}
+
+func newBatch(entries ...entry) *batch {
+	b := &batch{
+		streams: map[string]*logproto.Stream{},
+	}
+
+	for _, entry := range entries {
+		b.add(entry)
+	}
+
+	return b
+}
+
+func (b *batch) add(e entry) {
```

Comment:
I'm not sure how big these batches can be, but it's probably a good idea to limit them as soon as they cross a configurable size, maybe `128KB`?

Reply:
Looks like its easy, will try